### PR TITLE
Optimize ARRAY_UNION

### PR DESCRIPTION
**Pull request by yingsu00**
ARRAY_UNION can directly use the elementBlock in TypedSet as result.
This will reduce GC pressure and also improve performance. We also
provided specialization for Boolean type.

JMH benchmark before the change:
Benchmark                            (type)  Mode  Cnt   Score   Error  Units
BenchmarkArrayUnion.arrayUnion   BIGINT  avgt   20   43.703 ± 5.171  ns/op
BenchmarkArrayUnion.arrayUnion  VARCHAR  avgt   20   85.907 ± 3.745  ns/op
BenchmarkArrayUnion.arrayUnion   DOUBLE  avgt   20  113.791 ± 7.881  ns/op
BenchmarkArrayUnion.arrayIntersect  BOOLEAN  avgt   20  16.238 ± 0.359  ns/op

After:
Benchmark                       (type)  Mode  Cnt   Score   Error  Units
BenchmarkArrayUnion.arrayUnion   BIGINT  avgt   20  44.041 ± 5.132  ns/op
BenchmarkArrayUnion.arrayUnion  VARCHAR  avgt   20  75.538 ± 1.896  ns/op
BenchmarkArrayUnion.arrayUnion   DOUBLE  avgt   20  67.694 ± 6.582  ns/op
BenchmarkArrayUnion.arrayUnion  BOOLEAN  avgt   20  3.717 ± 0.282  ns/op

### CODE REVIEW

File: presto-main/src/main/java/com/facebook/presto/operator/aggregation/TypedSet.java [hunk 1]
```diff
@@ -118,6 +118,14 @@ public int positionOf(Block block, int position)
         return blockPositionByHash.get(getHashPositionOfElement(block, position));
     }
 
+    public Block getBlock()
+    {
+        if (containsNullElement) {
+            elementBlock.appendNull();
+        }
+        return elementBlock.build();
```
Comment (findepi) [inline presto-main/src/main/java/com/facebook/presto/operator/aggregation/TypedSet.java]: can i call `TypedSet.getBlock()` more than once?

File: presto-main/src/main/java/com/facebook/presto/operator/scalar/ArrayUnionFunction.java [hunk 2]
```diff
@@ -91,4 +103,28 @@ private static void appendBigintArray(Block array, AtomicBoolean containsNull, L
...
+        boolean[] results = new boolean[3];
+
+        for (int i = 0; i < block.getPositionCount(); i++) {
+            if (block.isNull(i)) {
+                results[0] = true;
+            }
+            else {
+                boolean value = BOOLEAN.getBoolean(block, i);
+                results[1] |= value;
+                results[2] |= !value;
```
Comment (findepi) [inline presto-main/src/main/java/com/facebook/presto/operator/scalar/ArrayUnionFunction.java]: Is this better than

```
else if (BOOLEAN.getBoolean(block, i)) {
    results[1] = true;
}
else {
    results[2] = true;
}
```

?

File: presto-main/src/main/java/com/facebook/presto/operator/scalar/ArrayUnionFunction.java [hunk 3]
```diff
@@ -91,4 +103,28 @@ private static void appendBigintArray(Block array, AtomicBoolean containsNull, L
...
     }
+
+    private static void appendTypedArray(Block array, TypedSet typedSet)
+    {
+        for (int i = 0; i < array.getPositionCount(); i++) {
+            typedSet.add(array, i);
+        }
+    }
+
+    private static boolean[] getDistinctBooleanValues(Block block)
```
Comment (findepi) [inline presto-main/src/main/java/com/facebook/presto/operator/scalar/ArrayUnionFunction.java]: I would return an instance:
```
private static class DistinctBooleanValues {
   private final boolean hasNull;
   private final boolean hasTrue;
   private final boolean hasFalse;
```

that would make usage on the calling site readable.

also, please move the method right below `booleanUnion`

File: presto-main/src/test/java/com/facebook/presto/operator/scalar/BenchmarkArrayDistinct.java [hunk 4]
```diff
@@ -69,8 +69,8 @@
 @BenchmarkMode(Mode.AverageTime)
 public class BenchmarkArrayDistinct
 {
-    private static final int POSITIONS = 100_000;
-    private static final int ARRAY_SIZE = 2;
+    private static final int POSITIONS = 10_000;
```
Comment (findepi) [inline presto-main/src/test/java/com/facebook/presto/operator/scalar/BenchmarkArrayDistinct.java]: why this change?

File: presto-main/src/main/java/com/facebook/presto/operator/scalar/ArrayUnionFunction.java [hunk 5]
```diff
@@ -91,4 +103,28 @@ private static void appendBigintArray(Block array, AtomicBoolean containsNull, L
             }
         }
     }
+
+    private static void appendTypedArray(Block array, TypedSet typedSet)
```
Comment (findepi) [inline presto-main/src/main/java/com/facebook/presto/operator/scalar/ArrayUnionFunction.java]: please move this back where it was, i.e. right below generic `union`

File: presto-main/src/main/java/com/facebook/presto/operator/scalar/ArrayUnionFunction.java (hunk 2 above)
Comment (martint) [reply presto-main/src/main/java/com/facebook/presto/operator/scalar/ArrayUnionFunction.java]: > Is this better than

The original form is certainly less readable, so unless there's a measurable and significant impact to performance, let's go with the readable version.

File: presto-main/src/main/java/com/facebook/presto/operator/scalar/ArrayUnionFunction.java [hunk 6]
```diff
@@ -75,6 +64,29 @@ public static Block bigintUnion(@SqlType("array(bigint)") Block leftArray, @SqlT
         return distinctElementBlockBuilder.build();
     }
 
+    @SqlType("array(boolean)")
+    public static Block booleanUnion(@SqlType("array(boolean)") Block leftArray, @SqlType("array(boolean)") Block rightArray)
```
Comment (martint) [inline presto-main/src/main/java/com/facebook/presto/operator/scalar/ArrayUnionFunction.java]: I'd write this whole thing as follows. No need to additional classes or containers.

```java
boolean hasNull = false;
boolean hasTrue = false;
boolean hasFalse = false;

for (int i = 0; i < left.getPositionCount(); i++) {
    if (left.isNull(i)) {
        hasNull = true;
    }
    else if (BOOLEAN.getBoolean(left, i)) {
        hasTrue = true;
    }
    else {
        hasFalse = true;
    }
}

for (int i = 0; i < right.getPositionCount(); i++) {
    if (right.isNull(i)) {
        hasNull = true;
    }
    else if (BOOLEAN.getBoolean(right, i)) {
        hasTrue = true;
    }
    else {
        hasFalse = true;
    }
}

if (hasNull) {
    distinctElementBlockBuilder.appendNull();
}
if (hasTrue) {
    BOOLEAN.writeBoolean(distinctElementBlockBuilder, true);
}
if (hasFalse) {
    BOOLEAN.writeBoolean(distinctElementBlockBuilder, false);
}

return distinctElementBlockBuilder.build();
```

File: presto-main/src/main/java/com/facebook/presto/operator/aggregation/TypedSet.java (hunk 1 above)
Comment (wenleix) [reply presto-main/src/main/java/com/facebook/presto/operator/aggregation/TypedSet.java]: Note once you called `getBlock`, the `TypedSet` becomes immutable.

That's why I am learning towards to generalize `JsonUtil#HashTable` for this common usage (i.e. constructing a Set over a block without copying the data), and deprecate `TypedSet` in the long term. 

We can discuss in person once @haozhun  is back.

Comment (wenleix) [reply presto-main/src/main/java/com/facebook/presto/operator/aggregation/TypedSet.java]: I opened an issue to discuss the solutions: https://github.com/prestodb/presto/issues/12033

File: presto-main/src/main/java/com/facebook/presto/operator/scalar/ArrayUnionFunction.java [hunk 7]
```diff
@@ -44,21 +45,9 @@ public static Block union(
...
-    {
-        for (int i = 0; i < array.getPositionCount(); i++) {
-            if (!typedSet.contains(array, i)) {
-                typedSet.add(array, i);
-                type.appendTo(array, i, blockBuilder);
-            }
-        }
+        appendTypedArray(leftArray, typedSet);
+        appendTypedArray(rightArray, typedSet);
+        return typedSet.getBlock();
```
Comment (wenleix) [inline presto-main/src/main/java/com/facebook/presto/operator/scalar/ArrayUnionFunction.java]: Note `ArrayUnion` should a cached `PageBuilder` to avoid allocate a new `BlockBuilder` in each invocation. 

This optimization cannot work with the cached `PageBuilder` solution.

Comment (wenleix) [inline presto-main/src/main/java/com/facebook/presto/operator/scalar/ArrayUnionFunction.java]: Note `ArrayUnion` should a cached `PageBuilder` to avoid allocate a new `BlockBuilder` in each invocation. 

This optimization cannot work with the cached `PageBuilder` solution. 

More discussions can be found in https://github.com/prestodb/presto/issues/12033

File: presto-main/src/main/java/com/facebook/presto/operator/scalar/ArrayUnionFunction.java (hunk 6 above)
Comment (yingsu00) [reply presto-main/src/main/java/com/facebook/presto/operator/scalar/ArrayUnionFunction.java]: @martint This was my originial writing except the part I removed the branching for true and false. Then I thought people don't like repeated code so come up with current version. I can change it back.

Comment (martint) [reply presto-main/src/main/java/com/facebook/presto/operator/scalar/ArrayUnionFunction.java]: Repeated code that's reasonably simple is not a problem. It has to be considered in the context of how much additional complexity and "conceptual weight" the alternatives introduce (new classes, interfaces, APIs, indirections, code that cannot be understood solely by looking at a localized area, etc)

Comment (findepi) [reply presto-main/src/main/java/com/facebook/presto/operator/scalar/ArrayUnionFunction.java]: If you want to avoid repeating the loops, you can always do this like this. But I think the unrolled version (@martint 's) is better.
```
for (Block block : ImmutableList.of(left, right)) {
   for (int i = 0; i < block.getPositionCount(); i++) {
       if (left.isNull(i)) {
           hasNull = true;
       ....
}
```

@martint in this context, would it make sense to break out of the `for` loop as soon as `hasNull`, `hasTrue` and `hasFalse` are all set?

Comment (martint) [reply presto-main/src/main/java/com/facebook/presto/operator/scalar/ArrayUnionFunction.java]: Very possibly, but it will depend on whether the extra checks are cheap on average compared to plowing through all the items. Some benchmarking would give us the answer.

Comment (findepi) [reply presto-main/src/main/java/com/facebook/presto/operator/scalar/ArrayUnionFunction.java]: And on the data -- for long arrays with mixed values it would be huge speedup. For long arrays without null, false or true, it will, by necessity, be slower. Benchmark would tell how much.

File: presto-main/src/main/java/com/facebook/presto/operator/scalar/ArrayUnionFunction.java (hunk 7 above)
Comment (yingsu00) [reply presto-main/src/main/java/com/facebook/presto/operator/scalar/ArrayUnionFunction.java]: As discussed in #11984 array_union needs the NoCopyTypedSet or JsonUtil#HashTable to support adding more than one blocks if we go through the first optimization. And we need to evaluate the performance using that solution. This PR shows the second optimization does show some gain for the array_union case. I'm waiting for people to chime in to decide which way to go, but I think we can keep both NoCopyTypedSet/ JsonUtil#HashTable and this modified TypedSet and the callers can decide which class to use.

File: presto-main/src/main/java/com/facebook/presto/operator/scalar/ArrayUnionFunction.java (hunk 5 above)
Comment (yingsu00) [reply presto-main/src/main/java/com/facebook/presto/operator/scalar/ArrayUnionFunction.java]: I organized the functions to be public methods first then private methods later. Do you think it's better to make each private helper method after its caller? Do we have a general guideline how to organize them?

File: presto-main/src/main/java/com/facebook/presto/operator/scalar/ArrayUnionFunction.java (hunk 2 above)
Comment (yingsu00) [reply presto-main/src/main/java/com/facebook/presto/operator/scalar/ArrayUnionFunction.java]: Using this way the performance is almost 3x worse.
Benchmark                        (type)  Mode  Cnt  Score   Error  Units
BenchmarkArrayUnion.arrayUnion  BOOLEAN  avgt   20  9.097 ± 0.496  ns/op

The branching in this case is on every value and is quite costly. It stalls the CPU pipeline and we should try to minimize it when possible.

File: presto-main/src/main/java/com/facebook/presto/operator/aggregation/TypedSet.java (hunk 1 above)
Comment (yingsu00) [reply presto-main/src/main/java/com/facebook/presto/operator/aggregation/TypedSet.java]: Right now it can't. But after TypedSet changed to use isDistinctFrom then we don't need to handle null differently, then this method can be called multiple times. Also all use cases of this method should only need to call it once, if needed.

File: presto-main/src/main/java/com/facebook/presto/operator/scalar/ArrayUnionFunction.java (hunk 2 above)
Comment (martint) [reply presto-main/src/main/java/com/facebook/presto/operator/scalar/ArrayUnionFunction.java]: I can't reproduce those results. These are the measurements I get for three variants I tried (on my macbook pro, 2.9 GHz Intel Core i9, Java 1.8.0_181):

1. The version I suggested above: 1.596 ± 0.053  ns/op
2. Similar to what I suggested above but using |= instead of if: 1.034 ± 0.052  ns/op
3. This PR: 2.740 ± 0.088  ns/op

How do you draw the conclusion that branching is stalling the CPU pipeline? Do you have a capture from perf that indicates so?

BTW, the benchmark in this PR is very biased. It doesn't test for different lengths of arrays, doesn't test when nulls are present, doesn't test for all true, all false, etc.

Comment (martint) [reply presto-main/src/main/java/com/facebook/presto/operator/scalar/ArrayUnionFunction.java]: You can make it even faster (0.782 ± 0.087 ns/op) by having a list of pre-computed blocks for all the 8 combinations of answers this method produces. Not that it matters much, especially for the additional complexity -- we're talking sub-nanosecond per element gains.

Comment (yingsu00) [reply presto-main/src/main/java/com/facebook/presto/operator/scalar/ArrayUnionFunction.java]: I reran the benchmark with 10, 100, 1000 elements in the array. The difference is reproducible if booleanUnion is calling getDistinctBooleanValues on my i7 processor on both JDK8 and 10

Calling getDistinctBooleanValues function, using |= (This PR)
Benchmark                       (ARRAY_SIZE)   (type)  Mode  Cnt    Score    Error  Units
BenchmarkArrayUnion.arrayUnion            10  BOOLEAN  avgt   20   10.823 ±  3.306  ms/op
BenchmarkArrayUnion.arrayUnion           100  BOOLEAN  avgt   20   28.701 ±  0.789  ms/op
BenchmarkArrayUnion.arrayUnion          1000  BOOLEAN  avgt   20  259.446 ± 64.891  ms/op

Calling getDistinctBooleanValues function, using if else
Benchmark                       (ARRAY_SIZE)   (type)  Mode  Cnt    Score    Error  Units
BenchmarkArrayUnion.arrayUnion            10  BOOLEAN  avgt   20   14.123 ±  0.441  ms/op
BenchmarkArrayUnion.arrayUnion           100  BOOLEAN  avgt   20   86.883 ±  4.807  ms/op
BenchmarkArrayUnion.arrayUnion          1000  BOOLEAN  avgt   20  869.110 ± 21.275  ms/op

The two getDistinctBooleanValues() functions are as follows and this is the only difference between the two runs:

```
private static boolean[] getDistinctBooleanValues(Block block)
    {
        boolean[] results = new boolean[3];

        for (int i = 0; i < block.getPositionCount(); i++) {
            if (block.isNull(i)) {
                results[0] = true;
            }
            else {
                boolean value = BOOLEAN.getBoolean(block, i);
                results[1] |= value;
                results[2] |= !value;
            }
        }
        return results;
    }
```

   ```
 private static boolean[] getDistinctBooleanValues(Block block)
    {
        boolean[] results = new boolean[3];

        for (int i = 0; i < block.getPositionCount(); i++) {
            if (block.isNull(i)) {
                results[0] = true;
            }
            else if (BOOLEAN.getBoolean(block, i)) {
                results[1] = true;
            }
            else {
                results[2] = true;
            }
        }
        return results;
    }
```
However as you said, writing the code in the inlined way as shown in your code below doesn't reproduce the performance difference.  

Originally I thought the difference shown in my above benchmark was because of the branch mis-prediction because the only difference between the two getDistinctBooleanValues implementation was the use of |=. But I looked at the assembly and didn't see any difference. In both cases the functions got inlined. So this may not be a branch misprediction issue. In this case the if-else are writing to two different variables instead of one and they tend to converge to true on the given workload, and the code block sizes are very small, so branch misprediction may not have such high impact. Also skylake has much better branch predictor which may also be the reason there was no difference on both i7 and i9 CPUs. Anyways, I get a profile on it to understand why there is the discrepancy in my runs.
